sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala

```diff
@@ -75,13 +78,28 @@ class InterpretedMutableProjection(expressions: Seq[Expression]) extends Mutable
     if (!e.nullable) {
       (v: Any) => writer(mutableRow, v)
     } else {
-      (v: Any) => {
-        if (v == null) {
-          mutableRow.setNullAt(i)
-        } else {
-          writer(mutableRow, v)
-        }
+      val nullSafeWriter: (InternalRow, Any) => Unit = e.dataType match {
+        case DecimalType.Fixed(precision, _) if precision > Decimal.MAX_LONG_DIGITS =>
```

Comment:
This won't handle the case where the decimal type is nested in a struct or array. But in that case, the target row won't be an unsafe row (since `InterpretedMutableProjection#target` would reject such a row, AFAICT), so the plain `writer` will be fine.